kvstore/sql_test.go

```diff
@@ -0,0 +1,298 @@
+	require.NoError(t, err)
+	defer db.Close()
+
+	s := SQL{Type: TypeEndorsement, DB: db}
+
+	emptyKey := ""
+
+	expectedErr := `the supplied key is empty`
+
+	err = s.Set(emptyKey, testVal)
+	assert.EqualError(t, err, expectedErr)
+}
+
+func TestSQL_Set_bad_val(t *testing.T) {
+	db, _, err := sqlmock.New()
+	require.NoError(t, err)
+	defer db.Close()
+
+	s := SQL{Type: TypeEndorsement, DB: db}
+
+	invalidJSON := ""
+
+	expectedErr := `the supplied val contains invalid JSON: unexpected end of JSON input`
+
+	err = s.Set(testKey, invalidJSON)
+	assert.EqualError(t, err, expectedErr)
+}
+
+func TestSQL_Set_db_layer_failure(t *testing.T) {
+	db, mock, err := sqlmock.New()
+	require.NoError(t, err)
+	defer db.Close()
+
+	s := SQL{Type: TypeEndorsement, DB: db}
+
+	dbErrorString := "a DB error"
+
+	e := mock.ExpectQuery(regexp.QuoteMeta("INSERT INTO endorsement(key, val) VALUES(?, ?)"))
+	e.WithArgs(testKey, testVal)
+	e.WillReturnError(errors.New(dbErrorString))
+
+	expectedErr := dbErrorString
+
+	err = s.Set(testKey, testVal)
+	assert.EqualError(t, err, expectedErr)
+
+	if err := mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("unfulfilled expectations: %s", err)
+	}
+}
+
+func TestSQL_Set_ok(t *testing.T) {
+	db, mock, err := sqlmock.New()
+	require.NoError(t, err)
+	defer db.Close()
+
+	s := SQL{Type: TypeEndorsement, DB: db}
+
+	e := mock.ExpectQuery(regexp.QuoteMeta("INSERT INTO endorsement(key, val) VALUES(?, ?)"))
+	e.WithArgs(testKey, testVal)
+	e.WillReturnRows()
+
+	err = s.Set(testKey, testVal)
+	assert.NoError(t, err)
+
+	if err := mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("unfulfilled expectations: %s", err)
+	}
+}
+
+func TestSQL_Del_empty_key(t *testing.T) {
+	db, _, err := sqlmock.New()
+	require.NoError(t, err)
+	defer db.Close()
+
+	s := SQL{Type: TypeEndorsement, DB: db}
+
+	emptyKey := ""
+
+	expectedErr := `the supplied key is empty`
+
+	err = s.Set(emptyKey, testVal)
+	assert.EqualError(t, err, expectedErr)
+}
+
+func TestSQL_Del_db_layer_failure(t *testing.T) {
+	db, mock, err := sqlmock.New()
+	require.NoError(t, err)
+	defer db.Close()
+
+	s := SQL{Type: TypeEndorsement, DB: db}
+
+	dbErrorString := "a DB error"
+
+	e := mock.ExpectQuery(regexp.QuoteMeta("DELETE FROM endorsement WHERE key = ?"))
+	e.WithArgs(testKey)
+	e.WillReturnError(errors.New(dbErrorString))
+
+	expectedErr := dbErrorString
+
+	err = s.Del(testKey)
+	assert.EqualError(t, err, expectedErr)
+
+	if err := mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("unfulfilled expectations: %s", err)
+	}
+}
+
+func TestSQL_Del_ok(t *testing.T) {
+	db, mock, err := sqlmock.New()
+	require.NoError(t, err)
+	defer db.Close()
+
+	s := SQL{Type: TypeEndorsement, DB: db}
+
+	e := mock.ExpectQuery(regexp.QuoteMeta("DELETE FROM endorsement WHERE key = ?"))
+	e.WithArgs(testKey)
+	e.WillReturnRows()
+
+	err = s.Del(testKey)
+	assert.NoError(t, err)
+
+	if err := mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("unfulfilled expectations: %s", err)
+	}
+}
```

Comment:
I do not see any test which tests TestSQLStore_Init_Close_cycle_ok ?
I see close() executed naturally, I understand, but where is s.Init(cfg) +ive case?

Reply:
oops, I forgot.  Will add.

Reply:
Fixed in 0dfbb0e